projects/adc_reading/src/main.c

```diff
@@ -0,0 +1,87 @@
+#include "adc.h"
+#include "delay.h"
+#include "gpio.h"
+#include "gpio_it.h"
+#include "interrupt.h"
+#include "interrupt_def.h"
+#include "log.h"
+#include "soft_timer.h"
+
+// Packages context for callback function outside main
+typedef struct Context_package {
+  GpioAddress button_addr;
+  InterruptSettings button_interrupt_settings;
+  AdcChannel adc_reading_channel;
+} Context_package;
+
+static void prv_interrupt_callback(const GpioAddress *address, void *context) {
+  // casting void* back for use outside main()
+  Context_package *p_package = context;
+
+  // Unpacking context package structure
+  GpioAddress button_addr = p_package->button_addr;
+  InterruptSettings button_interrupt_settings = p_package->button_interrupt_settings;
+  AdcChannel adc_reading_channel = p_package->adc_reading_channel;
+
+  // read and print adc reading data
+  uint16_t adc_reading_data = 0;
+  adc_read_converted(adc_reading_channel, &adc_reading_data);
+  LOG_DEBUG("ADC Reading Data: %d\n", adc_reading_data);
+
+  // restart interrupt watch
+  gpio_it_register_interrupt(&button_addr, &button_interrupt_settings, INTERRUPT_EDGE_FALLING,
+                             prv_interrupt_callback, p_package);
```

Comment:
You don't actually need to do this, the gpio interrupt never gets cleared. Just the `gpio_it_register_interrupt` in main will call this function whenever the interrupt occurs for the rest of the program (ie until the board is reflashed). (It's not like soft timers where if you want a loop you have to start a timer again.)

This means you don't need to pass around the `button_interrupt_settings` or the `button_addr` either.